examples/vnet/doca_gw_dpdk.c

```diff
@@ -0,0 +1,456 @@
+#include <stdio.h>
+#include "doca_gw_dpdk.h"
+#include "doca_log.h"
+
+DOCA_LOG_MODULE(doca_gw_dpdk);
+	
+#define DOCA_GET_SRC_IP(match, type) ((type == OUTER_MATCH) ? match->out_src_ip : match->in_src_ip) 
+#define DOCA_GET_DST_IP(match, type) ((type == OUTER_MATCH) ? match->out_dst_ip : match->in_dst_ip)
+#define DOCA_GET_SRC_PORT(match, type) ((type == OUTER_MATCH) ? match->out_src_port : match->in_src_port)
+#define DOCA_GET_DST_PORT(match, type) ((type == OUTER_MATCH) ? match->out_dst_port : match->in_dst_port)
+
+struct doca_gw_pipe_dpdk_flow_list pipe_flows;
+
+void doca_gw_init_dpdk(__rte_unused struct doca_gw_cfg *cfg)
+{
+	uint8_t idx;
+	struct doca_gw_pipe_dpdk_flow *pipe_flow;
+
+	LIST_INIT(&pipe_flows.free_head);
+	for(idx = 0 ; idx < MAX_PIP_FLOWS; idx++){
+		pipe_flow = &pipe_flows.pipe_flows[idx];
+		memset(pipe_flow, 0x0, sizeof(struct doca_gw_pipe_dpdk_flow));
+		pipe_flow->item_entry->item = &pipe_flow->items[idx];
+		LIST_INSERT_HEAD(&pipe_flows.free_head, pipe_flow, free_list);
+	}
+}
+
+/**
+ * @brief get free pipeline flow from arry by used flage, currently.
+ *    
+ * @entry match
+ *
+ * @return 
+ */
+static struct doca_gw_pipe_dpdk_flow*
+doca_gw_dpdk_get_free_pipe(void)
+{
+	struct doca_gw_pipe_dpdk_flow *pipe_flow = NULL;
+
+	if (!LIST_EMPTY(&pipe_flows.free_head)) {
+		pipe_flow = LIST_FIRST(&pipe_flows.free_head);
+		LIST_REMOVE(pipe_flow, free_list);
+		memset(pipe_flow, 0x0, sizeof(struct doca_gw_pipe_dpdk_flow));
+	}
+	return pipe_flow;
+}
+
+static void
+doca_gw_dpdk_pipe_release(struct doca_gw_pipe_dpdk_flow *pipe_flow)
+{
+	//TODO: need lock and set status 
+	LIST_INSERT_HEAD(&pipe_flows.free_head, pipe_flow, free_list);
+}
+
+static int
+doca_gw_dpdk_modify_eth_item(struct doca_dpdk_item_entry *entry,
+									struct doca_gw_match *match)
+{
+	struct rte_flow_item_eth *spec = &entry->item_data.eth.spec;
+
+	if ((entry->flags & DOCA_MODIFY_SMAC) && !doca_is_mac_zero(match->out_src_mac))
+		rte_ether_addr_copy((const struct rte_ether_addr*)match->out_src_mac, &spec->src);
+	if ((entry->flags & DOCA_MODIFY_DMAC) && !doca_is_mac_zero(match->out_dst_mac))
+		rte_ether_addr_copy((const struct rte_ether_addr*)match->out_dst_mac, &spec->dst);
+	return 0;
+}
+
+static uint16_t
+doca_gw_get_tunnel_eth_type(struct doca_gw_tun *tun)
+{
+	switch(tun->type) {
+		case DOCA_TUN_VXLAN:
+			return DOCA_ETH_P_IP;
```

Comment:
? you should test what is the type of the ip, and put here ipv4 or ipv6 (0x800)/0x86dd